feat: add SMTP auth & TLS support

dannykopping opened the pull request:
Closes https://github.com/coder/team-coconut/issues/1
Partially addresses https://github.com/coder/team-coconut/issues/5

Also changed the email template to use a hosted logo since [SVG support in email clients is still pitiful](https://css-tricks.com/a-guide-on-svg-support-in-email/).

dannykopping committed:
WIP: SMTP tests

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
WIP: TLS tests

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Flags

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping commented on codersdk/deployment.go:
```diff
@@ -2121,7 +2158,7 @@ Write out the current server config as YAML to stdout.`,
 			Default:     time.Minute.String(),
 			Group:       &deploymentGroupNotifications,
-			YAML:        "dispatch-timeout",
+			YAML:        "dispatchTimeout",
```
Drive-by: fixing previous mistakes of mine.

dannykopping committed:
Self-review

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
make lint/fmt

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Ignore typo in HELO

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
make update-flake / make update-golden-files

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
make lint / make gen

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Make tests more resilient across platforms

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Wait for server

Signed-off-by: Danny Kopping <danny@coder.com>

mtojek commented on .github/workflows/typos.toml:
```diff
@@ -20,6 +20,8 @@ hel = "hel"
 # typos doesn't like the EDE in TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA
 EDE = "EDE"
+# HELO is an SMTP command
+HELO = "HELO"
```
BTW is it HELO not EHLO?

dannykopping replied:
Both. If EHLO is not supported the library falls back to HELO.

mtojek commented on cli/testdata/server-config.yaml.golden:
```diff
@@ -493,13 +493,15 @@ userQuietHoursSchedule:
   # How long to wait while a notification is being sent before giving up.
   # (default: 1m0s, type: duration)
-  dispatch-timeout: 1m0s
+  dispatchTimeout: 1m0s
```
Looks like we're mixing _camelCase_ and _snake_case_ in the config. Should we stick with one format?

dannykopping replied:
Whoops, you're right. Thanks 👍 
We have a general problem with this it seems, but I'll at least try to make all of my additions internally consistent.

mtojek commented on coderd/notifications/dispatch/smtp.go:
```diff
@@ -133,47 +131,42 @@ func (s *SMTPHandler) dispatch(subject, htmlBody, plainBody, to string) Delivery
 			return false, xerrors.Errorf("'from' validation: %w", err)
 		}
-		err = c.Mail(from)
+		err = c.Mail(from, &smtp.MailOptions{})
```
Do you think we can build a fluent interface? Something like:

`c.Mail().Rcpts(...).Data(...).Send()`

dannykopping replied:
We could consider it, but I don't personally see the value. I'm very willing to be convinced though.

mtojek replied:
Readability, but I won't push for it. This syntax is also fine.

mtojek commented on coderd/notifications/dispatch/smtp.go:
```diff
@@ -93,37 +104,24 @@ func (s *SMTPHandler) dispatch(subject, htmlBody, plainBody, to string) Delivery
 
-		// Create an SMTP client.
-		c, err = smtp.NewClient(conn, smarthost)
+		// TODO: reuse client across dispatches (if possible).
```
I'm afraid we have to implement it now (or in follow-up as this PR is large). SMTP servers may have enabled restrictions limiting login attempts.

dannykopping replied:
I doubt it; Alertmanager does it this way and I'm not aware of any issues arising from that. I don't think we'll need this for the initial version.

mtojek replied:
I see, so it will just operate slower. Let's leave it then.

dannykopping replied:
Added https://github.com/coder/team-coconut/issues/15

mtojek reviewed (commented):
Have you done a trial with Google account?

mtojek commented on coderd/notifications/dispatch/smtp.go:
```diff
@@ -260,10 +253,214 @@ func (s *SMTPHandler) dispatch(subject, htmlBody, plainBody, to string) Delivery
+			}
+
+			// Warn that LOGIN is obsolete, but don't do it every time we dispatch a notification.
+			loginWarnOnce.Do(func() {
```
nit: Should we move it to initialization of the dispatcher?

dannykopping replied:
Good idea 👍

mtojek reviewed (approved):
👍

dannykopping committed:
Review comments

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
make gen

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Using hosted image URL rather than SVG since many clients do not support it

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping commented:
@mtojek heads up, added this after your review: https://github.com/coder/coder/pull/13902/commits/4f7a91024720d943be19c85dd7a30ed4494f77fe

Changed the email template to use a hosted logo since [SVG support in email clients is still pitiful](https://css-tricks.com/a-guide-on-svg-support-in-email/).

mtojek commented:
> Changed the email template to use a hosted logo since [SVG support in email clients is still pitiful](https://css-tricks.com/a-guide-on-svg-support-in-email/).

what about embedded PNG/base64 image?

dannykopping commented:
> > Changed the email template to use a hosted logo since [SVG support in email clients is still pitiful](https://css-tricks.com/a-guide-on-svg-support-in-email/).
> 
> what about embedded PNG/base64 image?

Tried that too, it shows like this:

<img width="1066" alt="image" src="https://github.com/user-attachments/assets/7ed7d220-b0f2-4f7c-800e-c871a77d7915">

mtojek commented:
Crap, that's bad. I'm wondering if we shouldn't host it on our site rather than on GitHub. I don't know if these raw links are stable API.

dannykopping committed:
Reference image hosted on coder.com instead

Relies on https://github.com/coder/coder.com/pull/518

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping commented:
I've created https://github.com/coder/coder.com/pull/518 and referenced the new URL in https://github.com/coder/coder/pull/13902/commits/5a8331f53016260787b29cd17e0c3d568783e7ff
Once that's merged I'll merge this one
